asllib/Typing.ml

```diff
@@ -2314,7 +2316,29 @@ module Annotate (C : ANNOTATE_CONFIG) : S = struct
     fun env v ldi -> add_constants v env ldi
   (* End *)
 
-  let rec annotate_stmt env s =
+  module ControlFlow = struct
+    type t = Nothing | Interrupt | Unreachable
+
+    let seq t1 t2 = if t1 = Nothing then t2 else t1
+
+    let guard env e t1 t2 =
+      match StaticModel.normalize_to_bool_opt env e with
+      | Some true -> t1
+      | Some false -> t2
+      | None -> (
+          match (t1, t2) with
+          | Nothing, _ | _, Nothing -> Nothing
+          | Unreachable, t | t, Unreachable ->
+              t (* Assertion that the condition always holds *)
+          | Interrupt, Interrupt -> Interrupt)
+
+    let check_interrupts ~loc name t () =
+      match t with
+      | Unreachable | Interrupt -> ()
+      | Nothing -> fatal_from loc (Error.NonReturningFunction name)
+  end
+
+  let rec annotate_stmt env s : stmt * env * ControlFlow.t =
```

Comment:
Do you have to perform the check while annotating statements?
Can't this check be done in a separate phase, after annotating the body of a subprogram?

Reply:
It's not necessarily easier to document, but I think it's cleaner in terms of separating concerns and making it easier to understand the static semantics.

Reply:
I think [2a3bb70](https://github.com/herd/herdtools7/commit/2a3bb709c941d89cd7fe7e1d0684303ad3f6c589) is really good. Please let's go with it. I left some comments.